### PR TITLE
lockstep refactoring of `KernelSetValue`

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- rewrite kernel with the lockstep programming model
- change hard coded chunk size to template parameter
- add documentation

[HowTo lockstep programming](http://picongpu.readthedocs.io/en/dev/prgpatterns/lockstep.html)

# Review
The kernel is now longer and is fully rewritten, please take a deeper look to the kernel.

# Tests
- [x] default 3D KHI

### CODE REVIEW

File: src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp [hunk 1]
```diff
@@ -82,20 +85,68 @@ getValue(T_Type& value)
...
+>
 struct KernelSetValue
 {
-    template <class DataBox, typename T_ValueType, typename Space>
-    DINLINE void operator()(DataBox data, const T_ValueType value, const Space size) const
+    /** set value to all elements
+     *
+     * @tparam T_DataBox PMacc::DataBox, type of the memory box
+     * @tparam T_ValueType type of the value
+     * @tparam T_SizeVecType PMacc::math::Vector, index type
```
Comment (ax3l) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: can you add for *what kind of index* this is? index in a block?

File: src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp [hunk 2]
```diff
@@ -82,20 +85,68 @@ getValue(T_Type& value)
...
+     *
+     * @param memBox box where all elements should be set to value
```
Comment (ax3l) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: box *of which* all elements *shall* be set to value

File: src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp [hunk 3]
```diff
@@ -174,30 +225,42 @@ class TaskSetValue<T_ValueType, T_dim, true> : public TaskSetValueBase<T_ValueTy
...
+        const DataSpace< dim > area_size( this->destination->getCurrentDataSpace( current_size ) );
 
-        if(area_size.productOfComponents() != 0)
+        if( area_size.productOfComponents() != 0 )
         {
             auto gridSize = area_size;
 
-            /* line wise thread blocks*/
-            gridSize.x() = ceil(double(gridSize.x()) / 256.);
+            // size of a chunk(in elements)
```
Comment (ax3l) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: add a space before `(`

File: src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp [hunk 4]
```diff
@@ -174,30 +225,42 @@ class TaskSetValue<T_ValueType, T_dim, true> : public TaskSetValueBase<T_ValueTy
 
     virtual void init()
     {
-        size_t current_size = this->destination->getCurrentSize();
-        const DataSpace<dim> area_size(this->destination->getCurrentDataSpace(current_size));
+        size_t current_size = this->destination->getCurrentSize( );
```
Comment (ax3l) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: can you add an inline doc-string to those two vars for clarity?

File: src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp [hunk 5]
```diff
@@ -174,30 +225,42 @@ class TaskSetValue<T_ValueType, T_dim, true> : public TaskSetValueBase<T_ValueTy
...
 
-        if(area_size.productOfComponents() != 0)
+        if( area_size.productOfComponents() != 0 )
         {
             auto gridSize = area_size;
 
-            /* line wise thread blocks*/
-            gridSize.x() = ceil(double(gridSize.x()) / 256.);
+            // size of a chunk(in elements)
+            constexpr int xChunkSize = 256;
```
Comment (ax3l) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: should be an unsigned type

Comment (ax3l) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: this constant might be a good candidate for `libPMacc/include/cudaSpecs.hpp`

File: src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp [hunk 6]
```diff
@@ -174,30 +225,42 @@ class TaskSetValue<T_ValueType, T_dim, true> : public TaskSetValueBase<T_ValueTy
...
+            constexpr uint32_t numWorkers = traits::GetNumWorkers<
+                xChunkSize
+            >::value;
 
-            auto destBox = this->destination->getDataBox();
+            // use chunks with 256 elements in x direction
```
Comment (ax3l) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: ... with *xChunkSize* elements in ...

File: src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp [hunk 7]
```diff
@@ -235,26 +298,45 @@ class TaskSetValue<T_ValueType, T_dim, false> : public TaskSetValueBase<T_ValueT
         {
             auto gridSize = area_size;
 
-            /* line wise thread blocks*/
-            gridSize.x()= ceil(double(gridSize.x()) / 256.);
+            // size of a chunk(in elements)
```
Comment (ax3l) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: add space before `(`

File: src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp [hunk 8]
```diff
@@ -235,26 +298,45 @@ class TaskSetValue<T_ValueType, T_dim, false> : public TaskSetValueBase<T_ValueT
...
+            constexpr int xChunkSize = 256;
```
Comment (ax3l) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: same as above (add to `libPMacc/include/cudaSpecs.hpp` ?)

File: src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp [hunk 9]
```diff
@@ -235,26 +298,45 @@ class TaskSetValue<T_ValueType, T_dim, false> : public TaskSetValueBase<T_ValueT
...
+            constexpr uint32_t numWorkers = traits::GetNumWorkers<
+                xChunkSize
+            >::value;
+
+            // use chunks with 256 elements in x direction
```
Comment (ax3l) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: same as above

**Comment by ax3l** [comment]
awww, the above review should have been submitted as `comment` not as `approve` :D

File: src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp [hunk 10]
```diff
@@ -82,20 +85,68 @@ getValue(T_Type& value)
...
+                uint32_t const
+            )
+            {
+                auto virtualWorkerIdx( SizeVecType::create( 0 ) );
+                virtualWorkerIdx.x( ) = linearIdx;
+
+                SizeVecType const idx( blockSize * blockIndex + virtualWorkerIdx );
+                if( idx.x() < size.x() )
+                    memBox( idx ) = taskSetValueHelper::getValue( value );
+            }
```
Comment (theZiz) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: Why is this so complex using vectors for obviously one dimensional adressing? What dimension has `idx`?

Comment (ax3l) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: this is just a C++ style array access `array( index )`

Comment (theZiz) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: So the complexity of the code is because of compatibility?

Comment (ax3l) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: ah you mean the full block, not the access in the line above...

Comment (theZiz) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: Yes. And it was probably more a libPMacc/PIConGPU related question than a real PR problem. ;)

File: src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp (hunk 1 above)
Comment (psychocoderHPC) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: This is defined for the parameter `size`, this is only the type with no domain relation

File: src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp (hunk 5 above)
Comment (psychocoderHPC) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: This value is not a CUDA magic number, it is the chunk size for the data without relation to CUDA.
The relation to CUDA is created by `traits::GetNumWorkers<xChunkSize>::value`

File: src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp (hunk 1 above)
Comment (ax3l) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: ok, just add "index type [for size]"

Comment (ax3l) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: ah, didn't see it's a parameter's template. sure!

File: src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp (hunk 10 above)
Comment (psychocoderHPC) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: We miss iterators in PMacc therefore we need to transform our indices in the user code.

File: src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp (hunk 9 above)
Comment (psychocoderHPC) [inline src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp]: I changed the comment

**Pull request merged**